File: leet/0069-sqrt/sqrt.go
func mySqrt(x int) int {
    
    if x < 2 {
        return x
    } else if x < 4 {
        return 1
    } else if x < 9 {
        return 2
    }
    
    a := x / 2
    for a * a > x { a = a / 2 }
    
    i := a
    for i * i <= x && i < 2 * a { i++ }
    
    return i - 1
}

File: leet/0069-sqrt/sqrt.js
/**
 * @param {number} x
 * @return {number}
 */
var mySqrt = function(x) {
    if (x < 2) return x

    let sqrt = function(s, e){
        while(s < e){
            let m = Math.floor((s + e) / 2)
            let p = m * m
            if (p == x){
                return m
            } else if(p < x){
                if ((m + 1) * (m + 1) > x) return m
                s = m + 1 
            } else{
                e = m - 1
            }
        }

        return s

    }

    return sqrt(1, x)
};
